File: pxr/usd/usdUtils/testenv/testUsdUtilsCoalescingDiagnosticDelegate.py
#!/pxrpythonsubst
#
# Copyright 2017 Pixar
#
# Licensed under the terms set forth in the LICENSE.txt file available at
# https://openusd.org/license.
#

from pxr import Tf, Sdf, UsdUtils

# Constructing an sdf path with "" gives a tf warning 
def _genWarning():
    Tf.Warn('.')

EXPECTED_LINE_NO   = 13
EXPECTED_FUNC_NAME = '__main__._genWarning'

delegate = UsdUtils.CoalescingDiagnosticDelegate()

# Test dumping of diagnostics (this will be compared in baseline/output.txt)
for i in range(0,10):
    _genWarning()

delegate.DumpCoalescedDiagnosticsToStdout()

# Test collection of unfiltered diagnostics
for i in range(0,3):
    _genWarning()

unfiltered = delegate.TakeUncoalescedDiagnostics()
assert len(unfiltered) == 3
for i in range(0,3):
    assert unfiltered[i].sourceLineNumber == EXPECTED_LINE_NO 
    assert unfiltered[i].sourceFunction == EXPECTED_FUNC_NAME

# Test collection of filtered diagnostics
for i in range(0,8):
    _genWarning()
filtered = delegate.TakeCoalescedDiagnostics()
assert len(filtered) == 1
filteredDiagnostic = filtered[0]

assert filteredDiagnostic.sharedItem.sourceLineNumber == EXPECTED_LINE_NO 
assert filteredDiagnostic.sharedItem.sourceFunction == EXPECTED_FUNC_NAME 

import pprint
pprint.pprint(filtered)
pprint.pprint(unfiltered)

for each in filtered:
    pprint.pprint(each)
    pprint.pprint(each.sharedItem)
    pprint.pprint(each.unsharedItems)
for each in unfiltered:
    pprint.pprint(each)

File: pxr/usd/usdUtils/testenv/testUsdUtilsCoalescingDiagnosticDelegate.cpp
//
// Copyright 2017 Pixar
//
// Licensed under the terms set forth in the LICENSE.txt file available at
// https://openusd.org/license.
//

#include "pxr/pxr.h"
#include "pxr/base/tf/diagnostic.h"
#include "pxr/usd/usdUtils/coalescingDiagnosticDelegate.h"

#include <iostream>
#include <cassert>
#include <set>

PXR_NAMESPACE_USING_DIRECTIVE

// Here we emit some warnings on the same line, so they 
// will get coalesced, and others on different lines, so they wont
// We expect to get 4 results when coalesced, 6 uncoalesced
static void 
EmitWarnings() {
    TF_WARN("aaaaaaaaaaaaaa"); TF_WARN("bbbbbbbbbbbbbb"); 
    TF_WARN("cccccccccccccc");               
    TF_WARN("dddddddddddddd");               
    TF_WARN("eeeeeeeeeeeeee"); TF_WARN("ffffffffffffff");
}

// The same helper, but for statuses
static void
EmitStatuses() {
    TF_STATUS("."); TF_STATUS("."); 
    TF_STATUS(".");               
    TF_STATUS(".");               
    TF_STATUS("."); TF_STATUS(".");
}

int main() {
    UsdUtilsCoalescingDiagnosticDelegate delegate;

    EmitWarnings();
    auto coalesced = delegate.TakeCoalescedDiagnostics();
    TF_AXIOM(coalesced.size() == 4);
    
    EmitWarnings();
    EmitStatuses();
    coalesced = delegate.TakeCoalescedDiagnostics();
    TF_AXIOM(coalesced.size() == 8);

    // ensure that the line numbers are unique
    std::set<size_t> sourceLineNumbers;
    for (const auto& p : coalesced) {
        sourceLineNumbers.insert(p.sharedItem.sourceLineNumber); 
    }
    TF_AXIOM(sourceLineNumbers.size() == 8); 

    EmitWarnings();
    EmitWarnings();
    auto unfiltered = delegate.TakeUncoalescedDiagnostics();
    TF_AXIOM(unfiltered.size() == 12);

    // ensure that the line numbers are not unique
    sourceLineNumbers.clear();
    for (const auto& i : unfiltered) {
        sourceLineNumbers.insert(i->GetSourceLineNumber());
    }
    TF_AXIOM(sourceLineNumbers.size() == 4); 

    std::cout << "-------------------------------------------\n";
    EmitWarnings();
    EmitWarnings();
    EmitStatuses();
    EmitStatuses();
    EmitWarnings();
    delegate.DumpCoalescedDiagnostics(std::cout);

    std::cout << "-------------------------------------------\n";

    EmitWarnings();
    EmitWarnings();
    EmitStatuses();
    EmitStatuses();
    EmitWarnings();
    delegate.DumpUncoalescedDiagnostics(std::cout);

    std::cout << "-------------------------------------------\n";
}
